### PR TITLE
Attributes module

### PR DESCRIPTION
**Pull request by pcattori**
# ↪️ Pull Request

Relates to #309 , #300 , #294 

Adds the `attributes` package containing:
- `tc.attribute` and `tc.Attribute`
- `tc.attribute_type` and `tc.AttributeType`
- `tc.subattribute` and `tc.SubAttribute`

Adds the `datasets` package containing:
- `tc.dataset`

The `attributes` and `datasets` packages are type checked via `mypy`.

Additional changes:
- Tests are now being linted by `flake8`
- `invoke` task for typechecking: `typecheck`
- CI performs typechecking
- Reference doc space for features in Beta

## 💻 Examples

```python
import tamr_unify_client as tc

dataset = ...
attr = tc.attribute.create(dataset, name='myattr', is_nullable=False, type=tc.attribute_type.String())

attrs = tc.dataset._attributes(dataset)
assert attr in attrs
```

## ✔️ PR Todo

- [x] Added/updated testing for this change
- [x] Included links to related issues/PRs
- [x] Update relevant [docs](https://github.com/Datatamer/tamr-client/tree/master/docs) + docstrings
- [x] Update the [CHANGELOG](https://github.com/Datatamer/tamr-client/blob/master/CHANGELOG.md) under the current `-dev` version:
  - Add changelog entries under any that apply: **BREAKING CHANGES**, **NEW FEATURES**, **BUG FIXES**.
  - Changelog entry format: `[#<issue number>](<link to issue>) <change description>`

### CODE REVIEW

File: tamr_unify_client/url.py [hunk 1]
```diff
@@ -0,0 +1,13 @@
+from dataclasses import dataclass
+from urllib.parse import urlparse
```
Comment (pcattori) [inline tamr_unify_client/url.py]: remove unused import

File: tamr_unify_client/dataset/resource.py [hunk 2]
```diff
@@ -58,6 +59,11 @@ def attributes(self):
         alias = self.api_path + "/attributes"
         return AttributeCollection(self.client, alias)
 
+    @property
+    def url(self) -> URL:
```
Comment (pcattori) [inline tamr_unify_client/dataset/resource.py]: need docstring

File: stubs/responses.pyi [hunk 3]
```diff
@@ -0,0 +1,19 @@
...
+def add(
+    method: Optional[str] = None,
+    url: Optional[str] = None,
+    status: Optional[int] = None,
+    json: Optional[JsonDict] = None,
+): ...
+
+T = TypeVar("T")
+
+def activate(T) -> T: ...
```
Comment (yizhiyintamr) [inline stubs/responses.pyi]: What does this def activate(T) function do?

Comment (pcattori) [reply stubs/responses.pyi]: Context: we are using the `responses` library in our testing suite. This 3rd party library does not come with type annotations, so we have [a couple options](https://mypy.readthedocs.io/en/stable/running_mypy.html#missing-imports) to deal with it:

1. We can simply tell `mypy` to ignore imports from `responses`.
2. We can write our own "stubs" to tell `mypy` what types to expect from the `responses` library.

(1) is easier but it means that `mypy` will let us pass anything to the `responses` library functions and won't complain.

(2) is a little fancier and means we have to write this `.pyi` file, but gives us nice type checking (e.g. `responses.add(asdf=False)` will complain that `asdf` is not a valid keyword to `responses.add` OR `responses.add(method=1)` will complain that `method` should be a `str`, not an `int`).

---

In the `responses` library, the `responses.activate` function is a [decorator](https://dbader.org/blog/python-decorators). Specifically, `responses.activate` takes a function as input and produce a function (with the same signature) as output.

We capture this knowledge here by saying, "for some arbitrary type `T`, the `activate` function should accept something of type `T` and produce something of type `T`".

Comment (yizhiyintamr) [reply stubs/responses.pyi]: Decorator article is a very good read. Thanks for the link. It makes sense now.

**Pull request merged**